lib/client/conntest/kube.go

```diff
@@ -0,0 +1,273 @@
+	defer cancelFunc()
+
+	restConfig := &rest.Config{
+		Host: "https://" + s.cfg.KubernetesPublicProxyHostPort,
+		TLSClientConfig: rest.TLSClientConfig{
+			CAData:   ca.TLSCA,
+			CertData: key.TLSCert,
+			KeyData:  key.PrivateKeyPEM(),
+		},
+	}
+
+	if s.cfg.TLSRoutingEnabled {
+		k8host := strings.Split(s.webProxyAddr, ":")[0]
+		// replace localhost with "127.0.0.1" so GetKubeTLSServerName can generate a domain kube.cluster.local.
+		if strings.EqualFold(k8host, "localhost") {
+			k8host = "127.0.0.1"
+		}
+		restConfig.TLSClientConfig.ServerName = client.GetKubeTLSServerName(k8host)
+		restConfig.Host = "https://" + s.webProxyAddr
+	}
+
+	client, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	_, err = client.CoreV1().Pods(req.KubernetesNamespace).List(ctxWithTimeout, v1.ListOptions{})
+	diag, diagErr = s.handleErrFromKube(ctxWithTimeout, connectionDiagnosticID, err, req.KubernetesNamespace)
+	if err != nil || diagErr != nil {
+		return diag, diagErr
+	}
+
+	traceType := types.ConnectionDiagnosticTrace_KUBE_PRINCIPAL
+	message := "Access to the Kubernetes Cluster granted."
+	connDiag, err := s.appendDiagnosticTrace(ctxWithTimeout, connectionDiagnosticID, traceType, message, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	connDiag.SetMessage(types.DiagnosticMessageSuccess)
+	connDiag.SetSuccess(true)
+
+	if err := s.cfg.UserClient.UpdateConnectionDiagnostic(ctxWithTimeout, connDiag); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return connDiag, nil
+}
+
+func (s KubeConnectionTester) handleErrFromKube(ctx context.Context, connectionDiagnosticID string, actionErr error, namespace string) (types.ConnectionDiagnostic, error) {
+	var kubeErr *kubeerrors.StatusError
+	if actionErr != nil && !errors.As(actionErr, &kubeErr) {
+		traceType := types.ConnectionDiagnosticTrace_UNKNOWN_ERROR
+		message := fmt.Sprintf("Unknown error. %v", actionErr)
+		connDiag, err := s.appendDiagnosticTrace(ctx, connectionDiagnosticID, traceType, message, actionErr)
+		return connDiag, trace.Wrap(err)
+	}
+
+	if kubeErr != nil && strings.Contains(kubeErr.ErrStatus.Message, "has no assigned groups or users") {
+		message := `User-associated roles do not configure "kubernetes_groups" or "kubernetes_users". Make sure that at least one is configured for the user.`
+		traceType := types.ConnectionDiagnosticTrace_RBAC_PRINCIPAL
+
+		connDiag, err := s.appendDiagnosticTrace(ctx, connectionDiagnosticID, traceType, message, actionErr)
+		return connDiag, trace.Wrap(err)
+	}
+
+	message := "User-associated roles define valid Kubernetes principals."
+	traceType := types.ConnectionDiagnosticTrace_RBAC_PRINCIPAL
+
+	_, err := s.appendDiagnosticTrace(ctx, connectionDiagnosticID, traceType, message, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if kubeErr != nil && (strings.Contains(kubeErr.ErrStatus.Message, "not found") || strings.Contains(kubeErr.ErrStatus.Message, "[00] access denied")) {
```

Comment:
To not have such long `if` statements that require you to read the entire thing to understand what is being checked, I'd do something more like

```go
if kubeErr != nil {
	isNotFound := strings.Contains(kubeErr.ErrStatus.Message, "not found")
	isAccessDenied := strings.Contains(kubeErr.ErrStatus.Message, "[00] access denied")

	if isNotFound || isAccessDenied {
		message := "You are not authorized to access this Kubernetes Cluster. Ensure your role grants access by adding it to the 'kubernetes_labels' property."
		traceType := types.ConnectionDiagnosticTrace_RBAC_KUBE
		connDiag, err := s.appendDiagnosticTrace(ctx, connectionDiagnosticID, traceType, message, actionErr)
		return connDiag, trace.Wrap(err)
	}

	cannotListPods := strings.Contains(kubeErr.ErrStatus.Message, "cannot list resource \"pods\"")
	if cannotListPods {
		message := fmt.Sprintf("You are not allowed to list pods in the %q namespace. "+
			"Make sure your \"kubernetes_groups\" or \"kubernetes_users\" exist in the cluster and grant you access to list pods.", namespace)
		traceType := types.ConnectionDiagnosticTrace_KUBE_PRINCIPAL
		connDiag, err := s.appendDiagnosticTrace(ctx, connectionDiagnosticID, traceType, message, actionErr)
		return connDiag, trace.Wrap(err)
	}

	// etc...
}
```

Thoughts?